### PR TITLE
Backup replication in golang

### PR DESCRIPTION
**Pull request by Dexus**
Remove all overhead for the replication in shell and make it small with golang.

@majewsky can you review the Go code and give some hints what can be done better if there is anything?

### CODE REVIEW

File: backup-tools/backup-replication-image/go-src/Gopkg.lock [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+  branch = "master"
+  name = "github.com/prometheus/procfs"
+  packages = [".","xfs"]
+  revision = "e645f4e5aaa8506fc71d6edbc5c4ff02c04c46f2"
+
+[[projects]]
+  branch = "backup-replication-golang"
+  name = "github.com/sapcc/containers"
+  packages = ["backup-tools/go-src/configuration","backup-tools/go-src/prometheus","backup-tools/go-src/swiftcli","backup-tools/go-src/utils"]
+  revision = "9793475af9c5a8a5689ea66be48c8907ccfae10b"
```
Comment (majewsky) [inline backup-tools/backup-replication-image/go-src/Gopkg.lock]: Don't vendor your own code. This is bound to cause a lot of unnecessary confusion.

File: backup-tools/backup-replication-image/go-src/Gopkg.toml [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+#  version = "2.4.0"
+
+
+[[constraint]]
+  branch = "master"
+  name = "github.com/ncw/swift"
+
+[[constraint]]
+  branch = "backup-replication-golang"
+  name = "github.com/sapcc/containers"
```
Comment (majewsky) [inline backup-tools/backup-replication-image/go-src/Gopkg.toml]: See above. Please remove this.

File: backup-tools/backup-replication-image/go-src/vendor/github.com/sapcc/containers/backup-tools/go-src/configuration/configuration.go [hunk 3]
```diff
@@ -5,9 +5,6 @@ import (
 	"os"
 	"path/filepath"
 	"strings"
-
-	"github.com/sapcc/containers/backup-tools/go-src/underscore"
-	"gopkg.in/yaml.v2"
```
Comment (Dexus) [inline backup-tools/backup-replication-image/go-src/vendor/github.com/sapcc/containers/backup-tools/go-src/configuration/configuration.go]: This will now break the yaml import that later need to run the replication...

File: backup-tools/go-src/configuration/configuration.go [hunk 4]
```diff
@@ -5,9 +5,6 @@ import (
 	"os"
 	"path/filepath"
 	"strings"
-
-	"github.com/sapcc/containers/backup-tools/go-src/underscore"
-	"gopkg.in/yaml.v2"
```
Comment (Dexus) [inline backup-tools/go-src/configuration/configuration.go]: This will now break the yaml import that later need to run the replication...

File: backup-tools/backup-restore/app.go [hunk 5]
```diff
@@ -14,11 +14,9 @@ import (
 
 	"github.com/sapcc/containers/backup-tools/go-src/configuration"
 	"github.com/sapcc/containers/backup-tools/go-src/swiftcli"
-	"github.com/sapcc/containers/backup-tools/go-src/underscore"
 	"github.com/sapcc/containers/backup-tools/go-src/utils"
 
 	"github.com/ncw/swift"
-	"gopkg.in/urfave/cli.v1"
```
Comment (Dexus) [inline backup-tools/backup-restore/app.go]: Why removing the cli?

File: backup-tools/backup-replication-image/go-src/job_queue.go [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+			uploadDone, upErr = swiftcli.SwiftUploadFile(to.SwiftCli, dlFilePath, &expiration, &fakeName)
+			if upErr != nil || !uploadDone {
+				PromGauge.SetError()
+				log.Printf("Worker%d File %d/%d (%s): Upload File Error: %s\n", id, j.FileNumber, j.FileAllCount, j.File, upErr)
+			}
+		}
+	}
+
+	// GoToEndWhileDownloadError used for exit the replication for error while downloading the file
+GoToEndWhileDownloadError:
```
Comment (majewsky) [inline backup-tools/backup-replication-image/go-src/job_queue.go]: You are always going here (there's no `return` before), so I'm pretty sure there's something wrong here in the success case. In general, please avoid `goto`.

File: backup-tools/backup-replication-image/go-src/job_queue.go [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+	EnvTo        []*Env
+	File         string
+	FileAllCount int
+	FileNumber   int
+}
+
+func doWork(id int, j Job) {
+	var dlErr, upErr error
+	var uploadDone bool
+	var dlFilePath string
```
Comment (majewsky) [inline backup-tools/backup-replication-image/go-src/job_queue.go]: It's bad style to declare variables at the start of the function. Declare them when you use them for the first time, with the `:=` operator, and take advantage of the variables going out of scope at the end of their block.

File: backup-tools/backup-replication-image/go-src/job_queue.go [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+	}
+	close(jobs)
+
+	// wait for workers to complete
+	wg.Wait()
+
+	os.RemoveAll(backupDir)
+
+	var replicationDuration = time.Since(startTime)
+	log.Printf("End replication task in %v\n", replicationDuration-(replicationDuration%time.Second))
```
Comment (majewsky) [inline backup-tools/backup-replication-image/go-src/job_queue.go]: Shorten this to `log.Printf("End replication task in %g seconds\n", time.Since(startTime).Seconds())`.

File: backup-tools/backup-replication-image/go-src/job_queue.go [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+		EnvFrom.Cfg.OsAuthVersion,
+		EnvFrom.Cfg.OsAuthURL,
+		EnvFrom.Cfg.OsUsername,
+		EnvFrom.Cfg.OsPassword,
+		EnvFrom.Cfg.OsUserDomainName,
+		EnvFrom.Cfg.OsProjectName,
+		EnvFrom.Cfg.OsProjectDomainName,
+		EnvFrom.Cfg.OsRegionName,
+		EnvFrom.Cfg.ContainerPrefix)
+	EnvFrom.Files, _ = swiftcli.SwiftListPrefixFiles(EnvFrom.SwiftCli, EnvFrom.Cfg.ContainerPrefix)
```
Comment (majewsky) [inline backup-tools/backup-replication-image/go-src/job_queue.go]: Errors should not be swallowed here (also down below).

File: backup-tools/backup-replication-image/go-src/job_queue.go [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+	}
+
+	// GoToEndWhileDownloadError used for exit the replication for error while downloading the file
+GoToEndWhileDownloadError:
+
+	if _, err := os.Stat(dlFilePath); !os.IsNotExist(err) {
+		os.Remove(dlFilePath)
+	}
+
+	currentFilesDone += 1
```
Comment (majewsky) [inline backup-tools/backup-replication-image/go-src/job_queue.go]: Since this is below the `GoToEndWhileDownloadError` label, does that mean you're counting a file as done even if transfer failed?

File: backup-tools/backup-replication-image/go-src/job_queue.go (hunk 6 above)
Comment (Dexus) [reply backup-tools/backup-replication-image/go-src/job_queue.go]: This is only in case the download was not success so you will for sure always going to this Mark but you over jump the upload of a not existing file.

File: backup-tools/backup-replication-image/go-src/job_queue.go (hunk 10 above)
Comment (Dexus) [reply backup-tools/backup-replication-image/go-src/job_queue.go]: Yes the variable name if not the best for that it is used.

File: backup-tools/backup-etcd-image/Dockerfile.debug [hunk 11]
```diff
@@ -1,11 +1,16 @@
...
-ADD files/backup-etcd.tar /bin/
-ADD files/backup-certs.tar /
-
+COPY --from=gobuilder /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/ca-certificates.crt
+COPY --from=gobuilder /go/bin/backup-etcd /replication-run
+COPY --from=gobuilder /backup/env /backup/env
 VOLUME ["/backup"]
-
-ENTRYPOINT ["/bin/backup-etcd"]
+ENTRYPOINT ["/backup-etcd"]
```
[the remark this answers is not in the document]
Comment (Dexus) [reply backup-tools/backup-etcd-image/Dockerfile.debug]: Support for debugging with busybox. Because the etcd image has no shell

File: backup-tools/backup-etcd-image/Dockerfile [hunk 12]
```diff
@@ -1,10 +1,15 @@
-FROM quay.io/coreos/etcd:v2.3.8
+FROM golang as gobuilder
+COPY go-src/ /go/src/github.com/sapcc/containers/backup-tools/backup-etcd-image/go-src/
+RUN set -uex; \
+    cd /go/src/github.com/sapcc/containers/backup-tools/backup-etcd-image/go-src; \
+    export GOPATH=/go; \
+    go get ./...; \
+    CGO_ENABLED=0 GOOS=linux go build -a -tags netgo -ldflags '-s -w' -installsuffix cgo -o /go/bin/backup-etcd
```
[the remark this answers is not in the document]
Comment (Dexus) [reply backup-tools/backup-etcd-image/Dockerfile]: 1. yes was not ever so while switch build on mac and linux
2. copy & paste; installsuffix I don't know why... the tags, are for the cross compiling never checked if that need for docker builds.
3. to make sure everything is there as a fallback if not
4. you are right thats not need here.

File: backup-tools/backup-replication-image/Dockerfile [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+COPY go-src/ /go/src/replication-run/
+
+RUN set -uex; \
+    cd /go/src/replication-run; \
+    export GOPATH=/go; \
+    go get ... ; \
+    CGO_ENABLED=0 GOOS=linux go build -a -tags netgo -ldflags '-s -w' -installsuffix cgo -o /go/bin/replication-run; \
+    mkdir -p /backup/env
+
+FROM scratch
```
[the remark this answers is not in the document]
Comment (Dexus) [reply backup-tools/backup-replication-image/Dockerfile]: Yes that's working. Never had an issue with build binary only images from scratch for micro services.

File: backup-tools/backup-restore/app.go (hunk 5 above)
Comment (majewsky) [reply backup-tools/backup-restore/app.go]: I did not remove this, goimports removed this automatically when I saved the file. Therefore, the import must have been unnecessary.

File: backup-tools/backup-replication-image/go-src/job_queue.go (hunk 7 above)
Comment (Dexus) [reply backup-tools/backup-replication-image/go-src/job_queue.go]: there is no alternative as to do it in this way. If there was a way to do so I would do it. Or what would you do there? each if is a own block and each for, switch so... where should I pre define it when not there for the loop and the if statements?

File: backup-tools/backup-replication-image/go-src/Gopkg.lock (hunk 1 above)
Comment (Dexus) [reply backup-tools/backup-replication-image/go-src/Gopkg.lock]: Sure but in a Docker-Build there is no other way... because it will else break the build while there is no .git directory and infos.

File: backup-tools/backup-replication-image/go-src/Gopkg.toml (hunk 2 above)
Comment (Dexus) [reply backup-tools/backup-replication-image/go-src/Gopkg.toml]: see above

File: backup-tools/backup-replication-image/go-src/job_queue.go (hunk 7 above)
Comment (majewsky) [reply backup-tools/backup-replication-image/go-src/job_queue.go]: This is what I can immediately refactor to declare variables more locally: https://gist.github.com/majewsky/7781e91d0921ced5cb5fe56cbc161f14

From there, you can refactor the function to get rid of the goto and the otherwise strange control flow. The logic would go something like this:

```go
if file replicated to all regions {
  return
}
path, err := download file
if err != nil {
  delete path
  complain
  return
}
foreach target region {
  upload file to region || complain
}
```

**Comment by Dexus** [comment]
@majewsky have done all steps, can you look again?
